File: 1-completion/javascript/completion.js
import OpenAI from 'openai';
import readline from 'node:readline';

const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] });

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const getUserMessage = () =>
  new Promise((resolve) => {
    rl.question('> ', (question) => {
      resolve(question);
    });
  });

const context = [];
context.push({ role: 'system', content: 'You are a useful assistant' });

///////////////////////////
// without streaming
///////////////////////////

while (true) {
  const question = await getUserMessage();

  if (question === 'bye') {
    process.exit();
  }

  context.push({ role: 'user', content: question });

  const response = await openai.chat.completions.create({
    model: 'gpt-4-1106-preview',
    temperature: 0.2,
    messages: context,
  });

  console.log(JSON.stringify(response, null, 2));

  const assistantResponse = response.choices[0].message.content;

  context.push({ role: 'assistant', content: assistantResponse });

  console.log(`\n🤖 ${assistantResponse}\n\n`);
  console.log(response.usage);
}

///////////////////////////
// with streaming
///////////////////////////

// while (true) {
//   const question = await getUserMessage();

//   if (question === 'bye') {
//     process.exit();
//   }

//   context.push({ role: 'user', content: question });

//   const stream = await openai.chat.completions.create({
//     model: 'gpt-4',
//     temperature: 0.2,
//     messages: context,
//     stream: true,
//   });

//   process.stdout.write('\n🤖 ');

//   for await (const chunk of stream) {
//     process.stdout.write(chunk.choices[0]?.delta?.content || '');
//   }

//   console.log('\n\n');
// }

File: 1-completion/python/completion.py
import os
import sys
from openai import OpenAI
from dotenv import load_dotenv, find_dotenv


_ = load_dotenv(find_dotenv())

openai = OpenAI(api_key=os.environ["OPENAI_API_KEY"])


context = []
context.append({"role": "system", "content": "You are a useful assistant."})


####################
# without streaming #
####################


while True:
    question = input("> ")

    if question == "bye":
        break

    context.append({"role": "user", "content": question})

    response = openai.chat.completions.create(
        model="gpt-4",
        temperature=0.2,
        messages=context,
    )

    assistantResponse = response.choices[0].message.content

    context.append({"role": "assistant", "content": assistantResponse})

    print("\n")

    print(f"🤖 {assistantResponse}\n\n")
    # print(response.usage)


##################
# with streaming #
##################


# while True:
#     question = input("> ")

#     if question == "bye":
#         break

#     context.append({"role": "user", "content": question})

#     stream = openai.chat.completions.create(
#         model="gpt-4",
#         temperature=0.2,
#         messages=context,
#         stream=True,
#     )

#     print("\n")
#     sys.stdout.write("🤖 ")

#     for part in stream:
#         chunk_message = part.choices[0].delta
#         if chunk_message and chunk_message.content:
#             sys.stdout.write(chunk_message.content)
#             sys.stdout.flush()

#     print("\n\n")
